http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java

```diff
@@ -221,4 +222,48 @@ private static void configureSslEngine(SSLEngine sslEngine) {
         sslParameters.setEndpointIdentificationAlgorithm("HTTPS");
         sslEngine.setSSLParameters(sslParameters);
     }
+
+    /**
+     * Create a {@link GenericFutureListener} that will propagate any failures or cancellations to the provided {@link Promise},
+     * or invoke the provided {@link Consumer} with the result of a successful operation completion. This is useful for chaining
+     * together multiple futures that may depend upon each other but that may not have the same return type.
+     * <p>
+     * Note that if you do not need the value returned by a successful completion (or if it returns {@link Void}) you may use
+     * {@link #runOrPropagate(Promise, Runnable)} instead.
+     *
+     * @param destination the Promise to notify upon failure or cancellation
+     * @param onSuccess   the Consumer to invoke upon success
+     */
+    public static <T> GenericFutureListener<Future<T>> consumeOrPropagate(Promise<?> destination, Consumer<T> onSuccess) {
+        return f -> {
+            if (f.isSuccess()) {
```

Comment:
> it's possible for a subsequent acquire to race against the release

Is it possible to write Junits to force this and test this scenarion ?

Reply:
Race conditions like this are difficult to unit-test, but let me add a unit test to ensure we await the release success.